kernel/src/process/process_vm/heap.rs

```diff
@@ -1,139 +1,166 @@
 // SPDX-License-Identifier: MPL-2.0
 
-use core::sync::atomic::{AtomicUsize, Ordering};
+use core::ops::Range;
 
 use align_ext::AlignExt;
 
 use crate::{
     prelude::*,
+    process::ResourceType,
+    util::random::getrandom,
     vm::{perms::VmPerms, vmar::Vmar},
 };
 
-/// The base address of user heap
-pub const USER_HEAP_BASE: Vaddr = 0x0000_0000_1000_0000;
-/// The max allowed size of user heap
-pub const USER_HEAP_SIZE_LIMIT: usize = 16 * 1024 * PAGE_SIZE; // 16 * 4MB
-
 #[derive(Debug)]
 pub struct Heap {
-    /// The lowest address of the heap
-    base: Vaddr,
-    /// The heap size limit
-    limit: usize,
-    /// The current heap highest address
-    current_program_break: AtomicUsize,
+    inner: RwMutex<Option<HeapInner>>,
+}
+
+#[derive(Clone, Debug)]
+struct HeapInner {
+    /// The data segment range of the process.
+    data_range: Range<Vaddr>,
```

Comment:
I don't think it's necessary to use a range here? We only need to know the size of the data segment.
```rust
    /// The size of all loaded data segments.
    data_size: usize,
```